src/Globals.cpp

```diff
@@ -249,227 +492,527 @@ static jstring JNICALL qore_exception_wrapper_get_message(JNIEnv*, jclass, jlong
-      }
+    {
+        const_cast<char*>("finalize0"),
+        const_cast<char*>("(J)V"),
+        reinterpret_cast<void*>(invocation_handler_finalize)
+    },
+    {
+        const_cast<char*>("invoke0"),
+        const_cast<char*>("(JLjava/lang/Object;Ljava/lang/reflect/Method;[Ljava/lang/Object;)Ljava/lang/Object;"),
+        reinterpret_cast<void*>(invocation_handler_invoke)
+    }
 };
 
-static JNINativeMethod qoreJavaApiNativeMethods[1] = {
-      {
-            const_cast<char*>("callFunction0"),
-            const_cast<char*>("(JLjava/lang/String;[Ljava/lang/Object;)Ljava/lang/Object;"),
-            reinterpret_cast<void*>(java_api_call_function)
-      },
+static JNINativeMethod qoreJavaApiNativeMethods[] = {
+    {
+        const_cast<char*>("callFunction0"),
+        const_cast<char*>("(JLjava/lang/String;[Ljava/lang/Object;)Ljava/lang/Object;"),
+        reinterpret_cast<void*>(java_api_call_function)
+    },
+    {
+        const_cast<char*>("callFunctionSave0"),
+        const_cast<char*>("(JLjava/lang/String;[Ljava/lang/Object;)Ljava/lang/Object;"),
+        reinterpret_cast<void*>(java_api_call_function_save)
+    },
+    {
+        const_cast<char*>("callStaticMethod0"),
+        const_cast<char*>("(JLjava/lang/String;Ljava/lang/String;[Ljava/lang/Object;)Ljava/lang/Object;"),
+        reinterpret_cast<void*>(java_api_call_static_method)
+    },
+    {
+        const_cast<char*>("callStaticMethodSave0"),
+        const_cast<char*>("(JLjava/lang/String;Ljava/lang/String;[Ljava/lang/Object;)Ljava/lang/Object;"),
+        reinterpret_cast<void*>(java_api_call_static_method_save)
+    },
+    {
+        // private native static QoreObject newObjectSave0(long pgm_ptr, String key, String class_name, Object...args);
+        const_cast<char*>("newObjectSave0"),
+        const_cast<char*>("(JLjava/lang/String;[Ljava/lang/Object;)Lorg/qore/jni/QoreObject;"),
+        reinterpret_cast<void*>(java_api_new_object_save)
+    },
 };
 
+#define NUM_QORE_JAVA_API_NATIVE_METHODS (sizeof(qoreJavaApiNativeMethods) / sizeof(JNINativeMethod))
+
 static JNINativeMethod qoreExceptionWrapperNativeMethods[2] = {
-      {
-            const_cast<char*>("finalize0"),
-            const_cast<char*>("(J)V"),
-            reinterpret_cast<void*>(qore_exception_wrapper_finalize)
-      },
-      {
-            const_cast<char*>("getMessage0"),
-            const_cast<char*>("(J)Ljava/lang/String;"),
-            reinterpret_cast<void*>(qore_exception_wrapper_get_message)
-      }
+    {
+        const_cast<char*>("finalize0"),
+        const_cast<char*>("(J)V"),
+        reinterpret_cast<void*>(qore_exception_wrapper_finalize)
+    },
+    {
+        const_cast<char*>("getMessage0"),
+        const_cast<char*>("(J)Ljava/lang/String;"),
+        reinterpret_cast<void*>(qore_exception_wrapper_get_message)
+    }
+};
+
+static JNINativeMethod qoreObjectNativeMethods[] = {
+    {
+        const_cast<char*>("className0"),
```

Comment:
`JNINativeMethod` needs a `char*` and string constants are `const char*`, so we need the cast to have correct code: http://xdprof.sourceforge.net/doxygen/structJNINativeMethod.html

Reply:
I don't think this makes the code correct. If the structure needs a `char *` then maybe it will destry it in the destructor, or maybe it will do some changes to it, etc. I know, that in this very example probably not, but it might (or with a future relase, etc). So casting the static string to char* is really not a good idea. The reason also can be smth like that Java will need the string to exists for particular amount of time and this way, they are telling you, that you cannot simply past some local read only variable, because it must preserve until jvm resources are released or so... 
So the correct code would probably look smth like a class that dynamically allocates the strings when the app is started and deallocates them in a destructor after jni module is done or so...
This way, all the global variables could be a members of this class, which can be a singleton for example, if it makes sense, and it can be called from a qorus (some hook to the init function etc).
I guess there is no such way now. So I probably merge this one, but I would like to address these issue rather sooner then later, otherwise we're going to add more and more code, that would need to be refactored one day...

Reply:
@sejvlond I assume that it was an error in the original JNI code - there is no documentation about this, so we can check the HS implementation and see if we had time, however the best proof of this hypothesis is to do a google search on all code examples using `RegisterNatives`; I haven't found a single one that uses anything other than `const char*` constants.  

Therefore if this would change to be something like you propose above, all existing JNI code would break.

You can do you own or check this one: https://stackoverflow.com/search?q=RegisterNatives

Anyway a lot of the issues that you have raised including this one are not part of the pull request.